### PR TITLE
Add tests on fixed ICEs

### LINKED ISSUE #29924: ICE: resolve_trait_associated_const: unexpected vtable type
the following code ices on nightly:

``` rust
#![feature(const_fn, associated_consts)]

trait Trait {
    const N: usize;
}

impl Trait {
    const fn n() -> usize { Self::N }
}
```

It should give an error instead noting that you can't do `Self::N`.

### LINKED ISSUE #38857: Locally pub but inaccessible items w/o stability attributes can cause an ICE
Spawned off of #38689

There are still some cases where referencing an inaccessible item can cause the compiler to ICE.

Here is one crafted example ([playpen](https://is.gd/WKNLvA))

```rust
fn main() {
    let a = std::sys::imp::process::process_common::StdioPipes { ..panic!() };
}
```

but if `std`'s module structure is revised then that example may fail to compile without ICE'ing for other reasons, so take care in understanding the issue: the change introduced by #38689 is skipping the requirement for stability attributes is slightly too narrow, in that it uses a local notion of pub/non-pub instead of overall accessibility.

One simple solution to this may be to just follow the suggestion here https://github.com/rust-lang/rust/pull/38689#issuecomment-270181214 and just remove the assertion entirely. (I did not want to adopt that solution for a PR to be backported to beta, but I think it is worth considering doing in nightly alone to resolve this low priority issue.)

### LINKED ISSUE #39665: ICE: "cannot relate bound region" when using conservative_impl_trait
[Playground](https://play.rust-lang.org/?gist=579ebe1a197a5f1db24990a41c7117a7&version=nightly&backtrace=0
)
```Rust
#![feature(conservative_impl_trait)]

fn batches(n: &u32) -> impl Iterator<Item=&u32> {
    std::iter::once(n)
}

fn main() {}
```

Output:
```
E:\mega3\projects\ice\ice>cargo build
   Compiling ice v0.1.0 (file:///E:/mega3/projects/ice/ice)
error: internal compiler error: C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\librustc\infer\region_inference/mod.rs:744: cannot relate bound region: '_#1r <= ReLateBound(DebruijnIndex { depth: 2 }, BrAnon(0))
 --> src\main.rs:3:24
  |
3 | fn batches(n: &u32) -> impl Iterator<Item=&u32> {
  |                        ^^^^^^^^^^^^^^^^^^^^^^^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\librustc_errors\lib.rs:376
stack backtrace:
   0:     0x7ffbf9a60d58 - std::panicking::Location::line::ha91c35bc276d0b51
   1:     0x7ffbf9a601e2 - std::panicking::Location::line::ha91c35bc276d0b51
   2:     0x7ffbf9a63c1d - std::panicking::rust_panic_with_hook::hf12525c539cafb95
   3:     0x7ffbf6d1288b - <unknown>
   4:     0x7ffbf6efbd99 - rustc::session::bug_fmt::h2b4624f40b53342d
   5:     0x7ffbf6efb7e2 - rustc::session::bug_fmt::h2b4624f40b53342d
   6:     0x7ffbf6efaf54 - rustc::session::bug_fmt::h2b4624f40b53342d
   7:     0x7ffbf6e378a4 - rustc::infer::region_inference::RegionVarBindings::make_subregion::h7d1c928b417a379e
   8:     0x7ffbf6e37177 - rustc::infer::region_inference::RegionVarBindings::make_eqregion::h7aba2537086b9605
   9:     0x7ffbf6e18347 - <rustc::infer::equate::Equate<'combine, 'infcx, 'gcx, 'tcx> as rustc::ty::relate::TypeRelation<'infcx, 'gcx, 'tcx>>::regions::h3e5abdb2851963c5
  10:     0x7ffbf6f66209 - rustc::ty::outlives::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::outlives_components::h006225a62a2484ea
  11:     0x7ffbf6e17a98 - <rustc::infer::equate::Equate<'combine, 'infcx, 'gcx, 'tcx> as rustc::ty::relate::TypeRelation<'infcx, 'gcx, 'tcx>>::tys::h0e345c52eceb6db5
  12:     0x7ffbf6e4ce89 - rustc::infer::InferCtxt::combine_fields::h8d78780cd9bc1f89
  13:     0x7ffbf6e537c4 - rustc::infer::InferCtxt::eq_types::h341a931021d55886
  14:     0x7ffbf6f15081 - rustc::traits::fulfill::GlobalFulfilledPredicates::check_duplicate_trait::h93cb52b260fb1258
  15:     0x7ffbf6e51365 - rustc::infer::InferCtxt::commit_from::h79f349c198c8b4f0
  16:     0x7ffbf6f12969 - <rustc::traits::fulfill::FulfillProcessor<'a, 'b, 'gcx, 'tcx> as rustc_data_structures::obligation_forest::ObligationProcessor>::process_obligation::h80789c9225e6da16
  17:     0x7ffbf6f104a4 - rustc::traits::fulfill::FulfillmentContext::take_deferred_obligations::h59157437f621c52e
  18:     0x7ffbf6f0fdb7 - rustc::traits::fulfill::FulfillmentContext::select_where_possible::hedb3ea6dfe818e1e
  19:     0x7ffbf79ef1a5 - rustc_typeck::check::FnCtxt::field_ty::h387f0899108f3c54
  20:     0x7ffbf79edfac - rustc_typeck::check::FnCtxt::field_ty::h387f0899108f3c54
  21:     0x7ffbf79dd506 - rustc_typeck::check::check_drop_impls::h632d17e4d693e43d
  22:     0x7ffbf79d9db0 - rustc_typeck::check::check_item_bodies::h9a3914a42542ad06
  23:     0x7ffbf7a5d7a0 - rustc_typeck::check_crate::h26811b192a872b84
  24:     0x7ffbf9e15c7e - rustc_driver::driver::count_nodes::h80b2ce48c34d37f0
  25:     0x7ffbf9d663c6 - <unknown>
  26:     0x7ffbf9df13f6 - rustc_driver::driver::compile_input::hc2e887bda4c69183
  27:     0x7ffbf9e42dfb - rustc_driver::run_compiler::h94efe000e99b91da
  28:     0x7ffbf9d1efbe - <unknown>
  29:     0x7ffbf9a66c51 - _rust_maybe_catch_panic
  30:     0x7ffbf9d4c806 - <unknown>
  31:     0x7ffbf9a5e1ce - std::sys::imp::thread::Thread::new::hd3ea3fbc2b0bf083
  32:     0x7ffc1abd8363 - BaseThreadInitThunk

error: Could not compile `ice`.

To learn more, run the command again with --verbose.
```

Other example (possibly different bug):

```Rust
#![feature(conservative_impl_trait)]

fn batches(n: &u32) -> impl Iterator<Item=&u32> {
    std::iter::empty()
}

fn main() {}
```
Output:
```
E:\mega3\projects\ice\ice>cargo build
   Compiling ice v0.1.0 (file:///E:/mega3/projects/ice/ice)
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'assertion failed: !obligation.predicate.has_escaping_regions()', C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\librustc\traits\select.rs:409
stack backtrace:
   0:     0x7ffbf9a60d58 - std::panicking::Location::line::ha91c35bc276d0b51
   1:     0x7ffbf9a601e2 - std::panicking::Location::line::ha91c35bc276d0b51
   2:     0x7ffbf9a63c1d - std::panicking::rust_panic_with_hook::hf12525c539cafb95
   3:     0x7ffbf6d12843 - <unknown>
   4:     0x7ffbf6f2236a - rustc::traits::select::SelectionContext::select::h1761016ce8ef2f1e
   5:     0x7ffbf6f1268d - <rustc::traits::fulfill::FulfillProcessor<'a, 'b, 'gcx, 'tcx> as rustc_data_structures::obligation_forest::ObligationProcessor>::process_obligation::h80789c9225e6da16
   6:     0x7ffbf6f104a4 - rustc::traits::fulfill::FulfillmentContext::take_deferred_obligations::h59157437f621c52e
   7:     0x7ffbf6f0fdb7 - rustc::traits::fulfill::FulfillmentContext::select_where_possible::hedb3ea6dfe818e1e
   8:     0x7ffbf783f1a5 - rustc_typeck::check::FnCtxt::field_ty::h387f0899108f3c54
   9:     0x7ffbf7838ea8 - rustc_typeck::check::FnCtxt::resolve_type_vars_with_obligations::h09ff96390f67faf0
  10:     0x7ffbf77de126 - rustc_typeck::check::demand::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, 'tcx>>::demand_coerce::h839b40a8b211aade
  11:     0x7ffbf782ebba - <rustc_typeck::check::GatherLocalsVisitor<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_pat::h8fcdbeae3dbc14f2
  12:     0x7ffbf782d4f3 - rustc_typeck::check::check_drop_impls::h632d17e4d693e43d
  13:     0x7ffbf7829db0 - rustc_typeck::check::check_item_bodies::h9a3914a42542ad06
  14:     0x7ffbf78ad7a0 - rustc_typeck::check_crate::h26811b192a872b84
  15:     0x7ffbf9e15c7e - rustc_driver::driver::count_nodes::h80b2ce48c34d37f0
  16:     0x7ffbf9d663c6 - <unknown>
  17:     0x7ffbf9df13f6 - rustc_driver::driver::compile_input::hc2e887bda4c69183
  18:     0x7ffbf9e42dfb - rustc_driver::run_compiler::h94efe000e99b91da
  19:     0x7ffbf9d1efbe - <unknown>
  20:     0x7ffbf9a66c51 - _rust_maybe_catch_panic
  21:     0x7ffbf9d4c806 - <unknown>
  22:     0x7ffbf9a5e1ce - std::sys::imp::thread::Thread::new::hd3ea3fbc2b0bf083
  23:     0x7ffc1abd8363 - BaseThreadInitThunk

error: Could not compile `ice`.

To learn more, run the command again with --verbose.
```

## Meta

```
rustc 1.17.0-nightly (c49d10207 2017-02-07)
binary: rustc
commit-hash: c49d10207a7e105525fb3bd71c18fde6fc2f5aed
commit-date: 2017-02-07
host: x86_64-pc-windows-msvc
release: 1.17.0-nightly
LLVM version: 3.9
```

### LINKED ISSUE #39872: Internal compiler error: cannot relate bound region (likely caused by `conservative_impl_trait`)
Related code:
```rust
pub struct PhysicalDevice<'a> {
	instance: &'a ::Instance<'a>,
	physical_device: VkPhysicalDevice,
	queue_families: Box<[QueueFamily]>
}

#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
pub struct QueueFamily {
	pub graphics: bool,
	pub compute: bool,
	pub transfer: bool,
	pub sparse_binding: bool
}

impl<'a> PhysicalDevice<'a> {
	// broken type signature
	pub fn queue_families(&self) -> impl Iterator<Item=&QueueFamily> {
		self.queue_families.iter()
	}
}
```

Compiler panic:
```
error: internal compiler error: C:\bot\slave\nightly-dist-rustc-win-msvc-64\build\src\librustc\infer\region_inference/mod.rs:744: cannot relate bound region: '_#4r <= ReLateBound(DebruijnIndex { depth: 2 }, BrAnon(0))
  --> galvan\src\instance\device/mod.rs:56:34
   |
56 |    pub fn queue_families(&self) -> impl Iterator<Item=&QueueFamily> {
   |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```

Note that everything is fine if I annotate the lifetimes:
```rust
// working type signature
pub fn queue_families(&'a self) -> impl Iterator<Item=&'a QueueFamily> {
	self.queue_families.iter()
}
```

I am running `rustc 1.17.0-nightly (ba7cf7cc5 2017-02-11)`, the MSVC x64 version.

### LINKED ISSUE #41210: thread 'rustc' panicked at 'Box<Any>'
I was in the middle of making a big changes to some code when this panic occurred. I was expecting a lot of compiler errors but got a compiler panic instead! :fearful: 

## Version Info

```
$ rustc --version --verbose
rustc 1.16.0 (30cf806ef 2017-03-10)
binary: rustc
commit-hash: 30cf806ef8881c41821fbd43e5cf3699c5290c16
commit-date: 2017-03-10
host: x86_64-unknown-linux-gnu
release: 1.16.0
LLVM version: 3.9
$ cargo --version
cargo-0.17.0-nightly (f9e5481 2017-03-03)
```

I ran the build with `cargo +nightly test` and the panic still occurred.

## Compiler Output

Run `cargo test` to get this.

Here's the full output with `RUST_BACKTRACE=1`:

```
error[E0412]: cannot find type `ArraySize` in this scope
  --> src/operations/item_type.rs:19:22
   |
19 |         size: Option<ArraySize>,
   |                      ^^^^^^^^^ not found in this scope
   |
   = help: possible candidate is found in another module, you can import it into scope:
             `use operations::scope::ArraySize;`

error[E0412]: cannot find type `TypeId` in this scope
  --> src/core/io/stdio.rs:16:50
   |
16 | fn define_stdin(scope: &mut ScopeStack, u8_type: TypeId) {
   |                                                  ^^^^^^ not found in this scope
   |
   = help: possible candidates are found in other modules, you can import them into scope:
             `use operations::scope::TypeId;`
             `use std::any::TypeId;`

error[E0412]: cannot find type `ArraySize` in this scope
  --> src/operations/item_type.rs:19:22
   |
19 |         size: Option<ArraySize>,
   |                      ^^^^^^^^^ not found in this scope
   |
   = help: possible candidate is found in another module, you can import it into scope:
             `use operations::scope::ArraySize;`

error[E0412]: cannot find type `TypeId` in this scope
  --> src/core/io/stdio.rs:16:50
   |
16 | fn define_stdin(scope: &mut ScopeStack, u8_type: TypeId) {
   |                                                  ^^^^^^ not found in this scope
   |
   = help: possible candidates are found in other modules, you can import them into scope:
             `use operations::scope::TypeId;`
             `use std::any::TypeId;`

error: no associated item named `Type` found for type `operations::scope::ScopeItem` in the current scope
  --> src/operations/type_definition.rs:15:13
   |
15 |             ScopeItem::Type(id) => Ok(id),
   |             ^^^^^^^^^^^^^^^^^^^

error[E0223]: ambiguous associated type
  --> src/operations/type_definition.rs:16:13
   |
16 |             ScopeItem::BuiltInFunction {id, ..} => Ok(id),
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous associated type
   |
   = note: specify the type using the syntax `<operations::scope::ScopeItem as Trait>::BuiltInFunction`

error: internal compiler error: /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/librustc/middle/mem_categorization.rs:1175: tuple struct pattern didn't resolve to variant or struct Err
  --> src/operations/type_definition.rs:15:13
   |
15 |             ScopeItem::Type(id) => Ok(id),
   |             ^^^^^^^^^^^^^^^^^^^

error: no associated item named `Type` found for type `operations::scope::ScopeItem` in the current scope
  --> src/operations/type_definition.rs:15:13
   |
15 |             ScopeItem::Type(id) => Ok(id),
   |             ^^^^^^^^^^^^^^^^^^^

error[E0223]: ambiguous associated type
  --> src/operations/type_definition.rs:16:13
   |
16 |             ScopeItem::BuiltInFunction {id, ..} => Ok(id),
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^ ambiguous associated type
   |
   = note: specify the type using the syntax `<operations::scope::ScopeItem as Trait>::BuiltInFunction`

error: internal compiler error: /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/librustc/middle/mem_categorization.rs:1175: tuple struct pattern didn't resolve to variant or struct Err
  --> src/operations/type_definition.rs:15:13
   |
15 |             ScopeItem::Type(id) => Ok(id),
   |             ^^^^^^^^^^^^^^^^^^^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/librustc_errors/lib.rs:375
stack backtrace:
   1:     0x7f862943740c - std::sys::imp::backtrace::tracing::imp::write::hf33ae72d0baa11ed
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:42
   2:     0x7f86294459ae - std::panicking::default_hook::{{closure}}::h59672b733cc6a455
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:351
   3:     0x7f8629445553 - std::panicking::default_hook::h1670459d2f3f8843
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:361
   4:     0x7f8629445e4b - std::panicking::rust_panic_with_hook::hcf0ddb069e7beee7
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:555
   5:     0x7f86263b79f8 - std::panicking::begin_panic::h7615035f91af0a46
   6:     0x7f8626568165 - rustc::session::opt_span_bug_fmt::{{closure}}::hb8dfe55fdd29ba42
   7:     0x7f8626567c65 - rustc::session::opt_span_bug_fmt::h0cd2e09801797662
   8:     0x7f8626567af4 - rustc::session::span_bug_fmt::hd9d57beec0176564
   9:     0x7f8626522e05 - rustc::middle::mem_categorization::MemCategorizationContext::cat_pattern_::hf9ee6cc28f4950bf
  10:     0x7f8626502fe8 - rustc::middle::expr_use_visitor::ExprUseVisitor::walk_expr::h67cf415f4e516db4
  11:     0x7f8626501e41 - rustc::middle::expr_use_visitor::ExprUseVisitor::consume_body::hf960393f75684590
  12:     0x7f8627e09a1c - <rustc_typeck::check::upvar::AdjustBorrowKind<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_fn::hcfd8923185beae25
  13:     0x7f8627da9267 - rustc::hir::intravisit::walk_expr::hb6d0c21c465f62f9
  14:     0x7f8627da8c1e - rustc::hir::intravisit::walk_expr::hb6d0c21c465f62f9
  15:     0x7f8627da8f4e - rustc::hir::intravisit::walk_expr::hb6d0c21c465f62f9
  16:     0x7f8627e08210 - rustc_typeck::check::upvar::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, 'tcx>>::closure_analyze::h260e6c5912cc14b4
  17:     0x7f8627e386a6 - rustc_typeck::check::check_bare_fn::h9f0916104d636ee0
  18:     0x7f8627e35875 - rustc_typeck::check::check_item_bodies::hd9e8c6a8692f314b
  19:     0x7f8627ea3497 - rustc_typeck::check_crate::h8c159f70d1009b33
  20:     0x7f86297e3cd7 - rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}::h040eea107b0d299b
  21:     0x7f86297e06c6 - rustc_driver::driver::phase_3_run_analysis_passes::haec055f3322a38aa
  22:     0x7f86297c4150 - rustc_driver::driver::compile_input::hd9f060ee16a643fb
  23:     0x7f862980e844 - rustc_driver::run_compiler::h762802568c0e140e
  24:     0x7f862971aedb - std::panicking::try::do_call::h935e2f773deaf841
  25:     0x7f862944ec8a - __rust_maybe_catch_panic
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libpanic_unwind/lib.rs:98
  26:     0x7f8629743112 - <F as alloc::boxed::FnBox<A>>::call_box::he43811d1f6894655
  27:     0x7f8629444804 - std::sys::imp::thread::Thread::new::thread_start::he668872ac11287ba
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/liballoc/boxed.rs:624
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/sys_common/thread.rs:21
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/sys/unix/thread.rs:84
  28:     0x7f8621efe6b9 - start_thread
  29:     0x7f86290fa82c - clone
  30:                0x0 - <unknown>

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/librustc_errors/lib.rs:375
stack backtrace:
   1:     0x7ff9adb2440c - std::sys::imp::backtrace::tracing::imp::write::hf33ae72d0baa11ed
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:42
   2:     0x7ff9adb329ae - std::panicking::default_hook::{{closure}}::h59672b733cc6a455
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:351
   3:     0x7ff9adb32553 - std::panicking::default_hook::h1670459d2f3f8843
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:361
   4:     0x7ff9adb32e4b - std::panicking::rust_panic_with_hook::hcf0ddb069e7beee7
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/panicking.rs:555
   5:     0x7ff9aaaa49f8 - std::panicking::begin_panic::h7615035f91af0a46
   6:     0x7ff9aac55165 - rustc::session::opt_span_bug_fmt::{{closure}}::hb8dfe55fdd29ba42
   7:     0x7ff9aac54c65 - rustc::session::opt_span_bug_fmt::h0cd2e09801797662
   8:     0x7ff9aac54af4 - rustc::session::span_bug_fmt::hd9d57beec0176564
   9:     0x7ff9aac0fe05 - rustc::middle::mem_categorization::MemCategorizationContext::cat_pattern_::hf9ee6cc28f4950bf
  10:     0x7ff9aabeffe8 - rustc::middle::expr_use_visitor::ExprUseVisitor::walk_expr::h67cf415f4e516db4
  11:     0x7ff9aabeee41 - rustc::middle::expr_use_visitor::ExprUseVisitor::consume_body::hf960393f75684590
  12:     0x7ff9ac4f6a1c - <rustc_typeck::check::upvar::AdjustBorrowKind<'a, 'gcx, 'tcx> as rustc::hir::intravisit::Visitor<'gcx>>::visit_fn::hcfd8923185beae25
  13:     0x7ff9ac496267 - rustc::hir::intravisit::walk_expr::hb6d0c21c465f62f9
  14:     0x7ff9ac495c1e - rustc::hir::intravisit::walk_expr::hb6d0c21c465f62f9
  15:     0x7ff9ac495f4e - rustc::hir::intravisit::walk_expr::hb6d0c21c465f62f9
  16:     0x7ff9ac4f5210 - rustc_typeck::check::upvar::<impl rustc_typeck::check::FnCtxt<'a, 'gcx, 'tcx>>::closure_analyze::h260e6c5912cc14b4
  17:     0x7ff9ac5256a6 - rustc_typeck::check::check_bare_fn::h9f0916104d636ee0
  18:     0x7ff9ac522875 - rustc_typeck::check::check_item_bodies::hd9e8c6a8692f314b
  19:     0x7ff9ac590497 - rustc_typeck::check_crate::h8c159f70d1009b33
  20:     0x7ff9aded0cd7 - rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}::h040eea107b0d299b
  21:     0x7ff9adecd6c6 - rustc_driver::driver::phase_3_run_analysis_passes::haec055f3322a38aa
  22:     0x7ff9adeb1150 - rustc_driver::driver::compile_input::hd9f060ee16a643fb
  23:     0x7ff9adefb844 - rustc_driver::run_compiler::h762802568c0e140e
  24:     0x7ff9ade07edb - std::panicking::try::do_call::h935e2f773deaf841
  25:     0x7ff9adb3bc8a - __rust_maybe_catch_panic
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libpanic_unwind/lib.rs:98
  26:     0x7ff9ade30112 - <F as alloc::boxed::FnBox<A>>::call_box::he43811d1f6894655
  27:     0x7ff9adb31804 - std::sys::imp::thread::Thread::new::thread_start::he668872ac11287ba
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/liballoc/boxed.rs:624
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/sys_common/thread.rs:21
                        at /buildslave/rust-buildbot/slave/stable-dist-rustc-linux/build/src/libstd/sys/unix/thread.rs:84
  28:     0x7ff9a65eb6b9 - start_thread
  29:     0x7ff9ad7e782c - clone
  30:                0x0 - <unknown>

error: Could not compile `brain`.
Build failed, waiting for other jobs to finish...
error: build failed
```

Since I was in the middle of editing, there's no commit to point to, so I just attached my source code as is so you have something to debug. When you download it and run `cargo test`, the compiler should panic. The repo is called [brain-lang/brain](https://github.com/brain-lang/brain).

I don't know how to narrow this issue down, or else I would try to find a smaller example for you. Hope this is good enough.

[brain-rustc-panic.zip](https://github.com/rust-lang/rust/files/911716/brain-rustc-panic.zip)

### LINKED ISSUE #41880: Compiler panicked: Ty::fn_sig() called on non-fn type: [type error]
I was playing around with some code which can be found [here](https://is.gd/t8CHlr) and is attached below.

When I click on `Mir` output I get the following output:
```
rustc 1.19.0-nightly (d3abc80b3 2017-05-09)
error: no method named `iter` found for type `Iterate<{integer}, [closure@<anon>:30:24: 30:33]>` in the current scope
  --> <anon>:32:24
   |
32 |     println!("{:?}", a.iter().take(10).collect::<Vec<usize>>());
   |                        ^^^^
   |
   = help: items from traits can only be used if the trait is implemented and in scope; the following trait defines an item `iter`, perhaps you need to implement it:
   = help: candidate #1: `core::slice::SliceExt`

error: internal compiler error: /checkout/src/librustc/ty/sty.rs:1316: Ty::fn_sig() called on non-fn type: [type error]

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:428
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

I don't have a rust installation locally right now, so I cannot create the backtrace. I'll add it later today.

The code which triggers this:
```
fn iterate<T, F>(initial: T, f: F) -> Iterate<T, F> {
    Iterate {
        state: initial,
        f: f,
    }
}

pub struct Iterate<T, F> {
    state: T,
    f: F
}

impl<T: Clone, F> Iterator for Iterate<T, F> where F: Fn(&T) -> T {
    type Item = T;

    #[inline]
    fn next(&mut self) -> Option<Self::Item> {
        self.state = (self.f)(&self.state);
        Some(self.state.clone())
    }
    #[inline]
    fn size_hint(&self) -> (usize, Option<usize>) { (std::usize::MAX, None) }
}

fn main() {
    let a = iterate(0, |x| x+1);
    
    println!("{:?}", a.iter().take(10).collect::<Vec<usize>>());
}
```

### LINKED ISSUE #43483: ICE: resolve_trait_associated_const: unexpected vtable type
```rust
trait VecN {
    const DIM: usize;
}
trait Mat {
    type Row: VecN;
}

fn m<M: Mat>() {
    let x = M::Row::DIM;
}
fn main() {
}
```
([Playpen](https://play.rust-lang.org/?gist=4d3d9aaf239510ccab021a6767eed6f7&version=nightly)) fails on current nightly with
```
   Compiling playground v0.0.1 (file:///playground)
error: internal compiler error: /checkout/src/librustc_const_eval/eval.rs:552: resolve_trait_associated_const: unexpected vtable type

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.20.0-nightly (afe145d22 2017-07-23) running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:490:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

### PR DESCRIPTION
**Pull request by topecongiro**
Closes #29924. Closes #38857. Closes #39665. Closes #39872.
Closes #41210. Closes #41880. Closes #43483.

Note that compile-fail/E0599.rs is for #41210.

### CODE REVIEW

File: src/test/compile-fail/associated-const-in-trait.rs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// #29924
+
+#![feature(const_fn, associated_consts)]
+
+trait Trait {
+    const N: usize;
+}
+
+impl Trait {
+    //~^ ERROR E0038
```
Comment (petrochenkov) [inline src/test/compile-fail/associated-const-in-trait.rs]: Could you use actual error messages instead of codes outside of `E0___.rs` tests?

Comment (topecongiro) [reply src/test/compile-fail/associated-const-in-trait.rs]: Thank you for your review. Updated and rebased.

Comment (estebank) [reply src/test/compile-fail/associated-const-in-trait.rs]: @topecongiro could you do that change in all `compile-fail` files?

**Review by estebank — APPROVED** [review]

**Comment by estebank** [comment]
#39953 is a PR, removed from description.

File: src/test/compile-fail/associated-const-in-trait.rs (hunk 1 above)
Comment (topecongiro) [reply src/test/compile-fail/associated-const-in-trait.rs]: @estebank Update tests and reabased. Also fixed the issue number (#39953 -> #39553).

**Pull request merged**